test/box-luatest/space_gc_consumers_test.lua

```diff
@@ -29,3 +30,376 @@ g.test_upgrade = function(cg)
+        t.assert_equals(#box.info.gc().consumers, 1)
+        t.assert_equals(box.info.gc().consumers[1].vclock, vclock)
+    end, {saved_vclock})
+end
+
+g.test_gc_consumers_object_name = function(cg)
+    local consumers = cg.master:exec(function()
+        local luuid = require('uuid')
+        local map = setmetatable({}, {__serialize = 'map'})
+        local consumers = {}
+
+        local uuid = luuid.str()
+        box.space._gc_consumers:replace{uuid, map, {type = 'replica'}}
+        table.insert(consumers, 'replica ' .. uuid)
+
+        uuid = luuid.str()
+        box.space._gc_consumers:replace{uuid, map, {object = 'invalid'}}
+
+        uuid = luuid.str()
+        box.space._gc_consumers:replace{uuid, map, map}
+
+        return consumers
+    end)
+    cg.master:restart()
+    cg.master:exec(function(expected_consumers)
+        t.assert_equals(#box.info.gc().consumers, 1)
+        t.assert_equals({box.info.gc().consumers[1].name}, expected_consumers)
+    end, {consumers})
+end
+
+-- The test that checks that invalid consumers do not break Tarantool. Also
+-- the test do some simple checks that consumers' issues are logged.
+g.test_gc_consumers_invalid_consumer = function(cg)
+    cg.master:exec(function()
+        local luuid = require('uuid')
+        local map = setmetatable({}, {__serialize = 'map'})
+
+        local uuid = luuid.str()
+        box.space._gc_consumers:replace{uuid, map,
+            {invalid = true, type = 'replica'}}
+
+        local uuid = luuid.str()
+        box.space._gc_consumers:replace{uuid, map, {type = 'unknown'}}
+
+        uuid = luuid.str()
+        local vclock = {key = 10}
+        box.space._gc_consumers:replace{uuid, vclock, map}
+
+        uuid = luuid.str()
+        vclock = {[33] = 1}
+        box.space._gc_consumers:replace{uuid, vclock, map}
+
+        box.space._gc_consumers:replace{'abc', map, map}
+    end)
+    cg.master:restart()
+    cg.master:exec(function()
+        t.assert_equals(box.info.gc().consumers, {})
+    end)
+    t.assert(cg.master:grep_log('Error while recovering GC consumer'))
+    t.assert(cg.master:grep_log('Error while recovering GC consumer abc'))
+    t.assert(cg.master:grep_log('has invalid type'))
+    t.assert(cg.master:grep_log('Invalid vclock'))
+end
+
+-- Test that outdated consumer becomes active after restart
+-- It is an important case because all persistent consumers
+-- are updated asynchronously, so many of active consumers
+-- may become outdated after restart.
+g.test_gc_consumers_outdated_consumer = function(cg)
+    cg.master:exec(create_and_fill_space)
+    cg.master:exec(function()
+        local uuid = require('uuid')
+        local replica_uuid = uuid.str()
+        box.space._cluster:insert{10, replica_uuid}
+        local vclock = box.info.vclock
+        vclock[1] = 1
+        local opts = {type = 'replica'}
+        box.space._gc_consumers:replace{replica_uuid, vclock, opts}
+        t.assert_equals(box.info.gc().consumers, {})
```

Comment:
But does it mean, that if a consumer is actually outdated and isn't going to be updated anymore, it will just pin the xlogs forever / until the disk is full?

Reply:
Yes, that's so. Outdated consumer can appear only after disk is full, so:
1. It's extremely rare.
2. Replica that owns outdated consumer will receive `XlogGapError` and is likely to be rebootstrapped (so the consumer will be dropped).

We decided to keep the patch as simple as possible and not handle outdated consumers because we doubt it'll even be needed. This logic could be implemented in future if someone needs it.

P.S. When disk is full, we cannot simply delete consumer from persistent state since we cannot write to WAL. So this logic becomes a bit complicated - we need to delete outdated consumer asynchronously or wait until WAL GC will actually delete xlogs.